library/core/src/any.rs

```diff
@@ -83,72 +83,115 @@
+//! reference.
+//! * The Consumer extracts a value or reference to the requested type from the `Request` object
+//! wrapped in an `Option<T>`; in the case of `dyn Error` the aforementioned `request_ref` and `
+//! request_value` methods mean that `dyn Error` users don't have to deal with the `Request` type at
+//! all (but `Error` implementors do). The `None` case of the `Option` suggests only that the
+//! Producer cannot currently offer an instance of the requested type, not it can't or never will.
 //!
 //! ## Examples
 //!
+//! The best way to demonstrate this is using an example implementation of `Error`'s `provide` trait
+//! method:
+//!
 //! ```
-//! # #![feature(provide_any)]
-//! use std::any::{Provider, Demand, request_ref};
+//! #![feature(error_generic_member_access)]
+//! use core::fmt;
+//! use core::any::Request;
 //!
-//! // Definition of MyTrait, a data provider.
-//! trait MyTrait: Provider {
+//! #[derive(Debug)]
+//! enum MyLittleTeaPot {
+//!     Empty,
+//! }
+//!
+//! #[derive(Debug)]
+//! struct MyBacktrace {
 //!     // ...
 //! }
 //!
-//! // Methods on `MyTrait` trait objects.
-//! impl dyn MyTrait + '_ {
-//!     /// Get a reference to a field of the implementing struct.
-//!     pub fn get_context_by_ref<T: ?Sized + 'static>(&self) -> Option<&T> {
-//!         request_ref::<T>(self)
+//! impl MyBacktrace {
+//!     fn new() -> MyBacktrace {
+//!         // ...
+//!         # MyBacktrace {}
 //!     }
 //! }
 //!
-//! // Downstream implementation of `MyTrait` and `Provider`.
-//! # struct SomeConcreteType { some_string: String }
-//! impl MyTrait for SomeConcreteType {
+//! #[derive(Debug)]
+//! struct SourceError {
 //!     // ...
 //! }
 //!
-//! impl Provider for SomeConcreteType {
-//!     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
-//!         // Provide a string reference. We could provide multiple values with
-//!         // different types here.
-//!         demand.provide_ref::<String>(&self.some_string);
+//! impl fmt::Display for SourceError {
+//!     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+//!         write!(f, "Example Source Error")
 //!     }
 //! }
 //!
-//! // Downstream usage of `MyTrait`.
-//! fn use_my_trait(obj: &dyn MyTrait) {
-//!     // Request a &String from obj.
-//!     let _ = obj.get_context_by_ref::<String>().unwrap();
+//! impl std::error::Error for SourceError {}
+//!
+//! #[derive(Debug)]
+//! struct Error {
+//!     source: SourceError,
+//!     backtrace: MyBacktrace,
+//! }
+//!
+//! impl fmt::Display for Error {
+//!     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+//!         write!(f, "Example Error")
+//!     }
+//! }
+//!
+//! impl std::error::Error for Error {
+//!     fn provide<'a>(&'a self, request: &mut Request<'a>) {
+//!         request
+//!             .provide_ref::<MyBacktrace>(&self.backtrace)
+//!             .provide_ref::<dyn std::error::Error + 'static>(&self.source);
```

Comment:
Is this a pattern that we want to encourage? The source should be exposed via the `source` trait method, not `provide`.

Reply:
Hmm, good point. For what it's worth I've mostly been naively copying and pasting stuff here and there with modifications to support removing the Provider trait. Will remove this and other references to source in the examples for `Request`.